### PR TITLE
Don't allocate for ResponseCookiesFeature

### LINKED ISSUE #676: ResponseCookiesFeature Allocates when unused
response-cookie-allocation

Resolved by: https://github.com/aspnet/HttpAbstractions/pull/675

### PR DESCRIPTION
**Pull request by benaadams**
* Don't allocate ResponseCookiesFeature when not used
* Don't allocate StringBuilder pool
* Don't allocate StringBuilders
* Don't allocate extra strings when constructing cookie string

Resolves #676

/cc @davidfowl @muratg @Tratcher @pakrym

### CODE REVIEW

File: src/Microsoft.AspNetCore.Http/Features/ResponseCookiesFeature.cs [hunk 1]
```diff
@@ -27,27 +24,13 @@ public class ResponseCookiesFeature : IResponseCookiesFeature
...
-
-        /// <summary>
-        /// Initializes a new <see cref="ResponseCookiesFeature"/> instance.
-        /// </summary>
-        /// <param name="features">
-        /// <see cref="IFeatureCollection"/> containing all defined features, including this
-        /// <see cref="IResponseCookiesFeature"/> and the <see cref="IHttpResponseFeature"/>.
-        /// </param>
-        /// <param name="builderPool">The <see cref="ObjectPool{T}"/>, if available.</param>
-        public ResponseCookiesFeature(IFeatureCollection features, ObjectPool<StringBuilder> builderPool)
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Http/Features/ResponseCookiesFeature.cs]: Side note: we can't remove public API (even if we change the behavior).

Comment (benaadams) [inline src/Microsoft.AspNetCore.Http/Features/ResponseCookiesFeature.cs]: Restored. Do I need to add back `AppendToStringBuilder` on `SetCookieHeaderValue`?

Comment (Tratcher) [inline src/Microsoft.AspNetCore.Http/Features/ResponseCookiesFeature.cs]: Yes, it was public in v1.0

Comment (benaadams) [inline src/Microsoft.AspNetCore.Http/Features/ResponseCookiesFeature.cs]: Restored

**Comment by benaadams** [comment]
Could also use [string.Concat](https://github.com/dotnet/coreclr/blob/master/src/mscorlib/src/System/String.Manipulation.cs#L370-L437) with a single `string[]` which would only be an allocation for the array as its quite efficient memory-wise.

Continually appending to the string is obv problematic, but the cookie is both known size and a known number of strings.

**Comment by Tratcher** [comment]
The string.concat option sounds like it's worth a try.

**Comment by pakrym** [comment]
Would be 2 allocations, 1 for array another for string, I'm thinking about extracting inplace string formatting into a struct, it would look like

```
var isf = new InplaceStringFormatter();
isf.AppendLength(s1);
isf.AppendLength(s2);
isf.AppendLength(s3);
isf.AppendValue(s1);
isf.AppendValue(s1);
isf.AppendValue(s1);
var s= isf.ToString();
```
Could reuse it in https://github.com/aspnet/HttpAbstractions/pull/716 too.

**Comment by pakrym** [comment]
@benaadams how do you feel about moving this PR to use https://github.com/aspnet/Common/pull/157 ?

**Comment by benaadams** [comment]
Yep

**Comment by benaadams** [comment]
Need to wait till stuff builds locally :-/

**Comment by pakrym** [comment]
@benaadams ping

**Comment by benaadams** [comment]
Still can't build
```
Errors in I:\Work\GitHub\HttpAbstractions\test\Microsoft.Net.Http.Headers.Tests\project.json
    Package Microsoft.Extensions.Testing.Abstractions 1.0.0-preview2-003121 is not compatible with netcoreapp1.0 (.NETCoreApp,Version=v1.0). Package Microsoft.Extensions.Testing.Abstractions 1.0.0-preview2-003121 supports:
      - net451 (.NETFramework,Version=v4.5.1)
      - netstandard1.6 (.NETStandard,Version=v1.6)
    One or more packages are incompatible with .NETCoreApp,Version=v1.0.
```
Etc

**Comment by benaadams** [comment]
Hmm, does that when I use ./build is ok in VS but can't run tests

**Comment by benaadams** [comment]
@pakrym updated

**Comment by pakrym** [comment]
You have old version of dotnet sdk somewhere, check what dotnet.exe gets used because 
```
    Package Microsoft.Extensions.Testing.Abstractions 1.0.0-preview2-003121 is not compatible with netcoreapp1.0 (.NETCoreApp,Version=v1.0). Package Microsoft.Extensions.Testing.Abstractions 1.0.0-preview2-003121 supports:
      - net451 (.NETFramework,Version=v4.5.1)
      - netstandard1.6 (.NETStandard,Version=v1.6)
```

was fixed long long time ago.

**Comment by pakrym** [comment]
https://github.com/aspnet/HttpAbstractions/commit/45696535048559188dce8b38a9fde82baa88893c

**Comment by khellang** [comment]
Is there a reason why the `(Pooled)HttpContextFactory` ctors taking an `ObjectPoolProvider` wasn't obsoleted with this PR?

**Pull request closed without merging**